src/goto-instrument/function_stubs.cpp

```diff
@@ -0,0 +1,276 @@
+    replace_return_value_symbol.insert(ret_val, return_value_symbol);
+  }
+
+  // Add the parameters as arguments to the call
+  for(const auto &parameter_identifier : type.parameter_identifiers())
+  {
+    const symbolt &parameter_symbol = ns.lookup(parameter_identifier);
+    internal_implementation_call.arguments().push_back(
+      parameter_symbol.symbol_expr());
+  }
+
+  goto_function.body.add(goto_programt::make_function_call(
+    code_function_callt(internal_implementation_call)));
+  // Maybe: Add location/comments maybe?
+
+  // Add the postcondition assumption
+  if(ensures.is_not_nil())
+  {
+    exprt ensures_replaced = ensures;
+    replace_return_value_symbol(ensures_replaced);
+    goto_function.body.add(goto_programt::make_assumption(
+      ensures_replaced, ensures.source_location()));
+  }
+
+  if(type.return_type() != empty_typet())
+  {
+    code_returnt return_expr(return_value_symbol);
+    goto_function.body.add(goto_programt::make_return(return_expr));
+  }
+
+  // Append the end of function instruction
+  goto_function.body.add(
+    goto_programt::make_end_function(ensures.source_location()));
+}
+
+bool function_stubst::operator()()
+{
+  // Get the body of the requested function to stub
+  goto_functiont function_body;
+  auto requested_function_body =
+    goto_functions.function_map.find(function_name);
+  if(requested_function_body == goto_functions.function_map.end())
+  {
+    log.error() << "The body of function: " << function_name
+                << "doesn't exist in the goto_model\n";
+    return true;
+  }
+  function_body.copy_from(requested_function_body->second);
+
+  log.debug() << "Stubbing function: " << function_name << "\n";
+
+  // Add a declaration of the impl function, that will be called in
+  // the function body. Create a new function symbol for the internal
+  // implementation function
+  if(!symbol_table.has_symbol(function_name))
+  {
+    log.error() << "Function: " << function_name
+                << "doesn't exist in the symbol table\n";
+    return true;
+  }
+  symbolt function_symbol = symbol_table.lookup_ref(function_name);
+  const irep_idt stub_function_name = stub_name_of_function(function_name);
+  function_symbol.name = stub_function_name;
+  function_symbol.location = source_locationt();
+
+  // Question: Maybe I should also change the symbol module?
+  function_symbol.base_name = stub_function_name;
+  function_symbol.pretty_name = stub_function_name;
+
+  if(symbol_table.has_symbol(function_symbol.name))
+  {
+    log.error() << "The internal stub implementation name: "
+                << function_symbol.name
+                << "already exists in the symbol table\n";
+    return true;
+  }
+  symbol_table.add(function_symbol);
+  stub_function(function_name, function_body);
+
+  // Delete all other functions and just keep the stub so that we can
+  // output it in C
+  goto_functions.clear();
```

Comment:
This seems like it should be done by our caller -- replacing a function with a stub doesn't seem like it should always mean clearing the goto-model

Reply:
Thanks, it is now done in the caller